Generate similar is a open source nlp library used for unsupervised topic modeling

Topic modelling is a concept in nlp its to find topics based on article or document.gensim is used in topic modeling.

Document is a form of statement or sentence.

corpus is a collection of documents,each document is seperated by , and with"" as 1 st doc,2nd doc,3rd doc etc.,



vector is a mathematical representation of a document.for 1 st doc we give some number and that number is a vector



Model is an algorithym used for transforming vectors from one representation to another .we are transforming from one vector to another vector.that is called model

prettyprint is used in gensim its like normal print in gensim we call it as prettyprint.


In [1]:
text_corpus = [
    "NASA’s Artemis program aims to return humans to the Moon. Scientists believe lunar exploration will pave the way for Mars missions. SpaceX also plays a key role in deep space transport.",
    "Artificial Intelligence is transforming industries. Machine learning algorithms power recommendation systems, autonomous vehicles, and chatbots. Ethical concerns around bias and data privacy remain important.",
    "Global temperatures are rising due to greenhouse gas emissions. Renewable energy sources like wind and solar are crucial to reduce dependence on fossil fuels. Sea levels continue to rise annually.",
    "A balanced diet and regular exercise improve physical and mental health. Doctors recommend at least 150 minutes of moderate activity per week. Hydration and sleep are also essential for well-being.",
    "Understanding budgeting, saving, and investing is key to financial independence. Compound interest can grow savings over time. Many people now use mobile apps to track expenses and build wealth.",
    "The history of computing spans from the abacus to quantum computers. Alan Turing is considered a pioneer in computer science. The invention of microprocessors revolutionized the tech industry."
    "NASA’s Artemis program aims to return humans to the Moon. Scientists believe lunar exploration will pave the way for Mars missions. SpaceX also plays a key role in deep space transport.",
    "Artificial Intelligence is transforming industries. Machine learning algorithms power recommendation systems, autonomous vehicles, and chatbots. Ethical concerns around bias and data privacy remain important.",
    "Global temperatures are rising due to greenhouse gas emissions. Renewable energy sources like wind and solar are crucial to reduce dependence on fossil fuels. Sea levels continue to rise annually.",
    "A balanced diet and regular exercise improve physical and mental health. Doctors recommend at least 150 minutes of moderate activity per week. Hydration and sleep are also essential for well-being.",
    "Understanding budgeting, saving, and investing is key to financial independence. Compound interest can grow savings over time. Many people now use mobile apps to track expenses and build wealth.",
    "The history of computing spans from the abacus to quantum computers. Alan Turing is considered a pioneer in computer science. The invention of microprocessors revolutionized the tech industry."
]







In [2]:
import pprint

# Correct stoplist splitting (by whitespace)
stoplist = set('for a of the and to in'.split())#set so no duplicates

# Create a list of lists with words filtered by stoplist
texts = [
    [word for word in document.lower().split() if word not in stoplist]
    for document in text_corpus
]

# Pretty print the result
pprint.pprint(texts)


[['nasa’s',
  'artemis',
  'program',
  'aims',
  'return',
  'humans',
  'moon.',
  'scientists',
  'believe',
  'lunar',
  'exploration',
  'will',
  'pave',
  'way',
  'mars',
  'missions.',
  'spacex',
  'also',
  'plays',
  'key',
  'role',
  'deep',
  'space',
  'transport.'],
 ['artificial',
  'intelligence',
  'is',
  'transforming',
  'industries.',
  'machine',
  'learning',
  'algorithms',
  'power',
  'recommendation',
  'systems,',
  'autonomous',
  'vehicles,',
  'chatbots.',
  'ethical',
  'concerns',
  'around',
  'bias',
  'data',
  'privacy',
  'remain',
  'important.'],
 ['global',
  'temperatures',
  'are',
  'rising',
  'due',
  'greenhouse',
  'gas',
  'emissions.',
  'renewable',
  'energy',
  'sources',
  'like',
  'wind',
  'solar',
  'are',
  'crucial',
  'reduce',
  'dependence',
  'on',
  'fossil',
  'fuels.',
  'sea',
  'levels',
  'continue',
  'rise',
  'annually.'],
 ['balanced',
  'diet',
  'regular',
  'exercise',
  'improve',
  'physical',
  'mental',

In [3]:
#now count and word frequencies
from collections import defaultdict
frequency=defaultdict(int)
for text in texts:
    for token in text:
        frequency[token]+=1

#only keep words that appear more than once
processed_corpus=[[token for token in text if frequency[token]>=2]for text in texts]
pprint.pprint(processed_corpus)

[['artemis',
  'program',
  'aims',
  'return',
  'humans',
  'moon.',
  'scientists',
  'believe',
  'lunar',
  'exploration',
  'will',
  'pave',
  'way',
  'mars',
  'missions.',
  'spacex',
  'also',
  'plays',
  'key',
  'role',
  'deep',
  'space',
  'transport.'],
 ['artificial',
  'intelligence',
  'is',
  'transforming',
  'industries.',
  'machine',
  'learning',
  'algorithms',
  'power',
  'recommendation',
  'systems,',
  'autonomous',
  'vehicles,',
  'chatbots.',
  'ethical',
  'concerns',
  'around',
  'bias',
  'data',
  'privacy',
  'remain',
  'important.'],
 ['global',
  'temperatures',
  'are',
  'rising',
  'due',
  'greenhouse',
  'gas',
  'emissions.',
  'renewable',
  'energy',
  'sources',
  'like',
  'wind',
  'solar',
  'are',
  'crucial',
  'reduce',
  'dependence',
  'on',
  'fossil',
  'fuels.',
  'sea',
  'levels',
  'continue',
  'rise',
  'annually.'],
 ['balanced',
  'diet',
  'regular',
  'exercise',
  'improve',
  'physical',
  'mental',
  'health.'

import defaultdict to keep count of word frequencies .then we instantiate defaultdict into frequency object.texts is the collection of words now where words are tokens or terms for every token in texts we take that text and load it into frequency object.if frequency of token>1 then that token will be taken and stored into processed_corpus above

create dictionary .dictionary is available in corpora during topic modeling we have dictionary  so we can refer dictionaryand convert it into some numbers.for that we import corpora from gensim.within corpora we have dictionaryso in dictionary we pass these processed_corpus

In [4]:
from gensim import corpora
dictionary=corpora.Dictionary(processed_corpus)
print(dictionary)


Dictionary<132 unique tokens: ['aims', 'also', 'artemis', 'believe', 'deep']...>


In [5]:
#vector reprsentation
pprint.pprint(dictionary.token2id)

{'150': 70,
 'abacus': 115,
 'activity': 71,
 'aims': 0,
 'alan': 116,
 'algorithms': 23,
 'also': 1,
 'annually.': 45,
 'apps': 92,
 'are': 46,
 'around': 24,
 'artemis': 2,
 'artificial': 25,
 'at': 72,
 'autonomous': 26,
 'balanced': 73,
 'believe': 3,
 'bias': 27,
 'budgeting,': 93,
 'build': 94,
 'can': 95,
 'chatbots.': 28,
 'compound': 96,
 'computer': 117,
 'computers.': 118,
 'computing': 119,
 'concerns': 29,
 'considered': 120,
 'continue': 47,
 'crucial': 48,
 'data': 30,
 'deep': 4,
 'dependence': 49,
 'diet': 74,
 'doctors': 75,
 'due': 50,
 'emissions.': 51,
 'energy': 52,
 'essential': 76,
 'ethical': 31,
 'exercise': 77,
 'expenses': 97,
 'exploration': 5,
 'financial': 98,
 'fossil': 53,
 'from': 121,
 'fuels.': 54,
 'gas': 55,
 'global': 56,
 'greenhouse': 57,
 'grow': 99,
 'health.': 78,
 'history': 122,
 'humans': 6,
 'hydration': 79,
 'important.': 32,
 'improve': 80,
 'independence.': 100,
 'industries.': 33,
 'intelligence': 34,
 'interest': 101,
 'invention': 1

vector representation is from one vector to another vector.we use stored values of 'dictionary' object we take all tokens from dictionary values which contains 132 unique tokens and convert these tokens into id.so use cmd token2id above

in tuple below 1st occurrence is id and 2nd occurrence is count.in above all are assigned with ids.


In [7]:
new_doc="human computer interaction"
new_vec=dictionary.doc2bow(new_doc.lower().split())
print(new_vec)

[(117, 1)]


1st occurrence is id and 2nd occurrence is count here count is the vector

doc2bow is document to bag of words.here we used the dictionary which we created.count is how many times the id of word is occurring.here word with id 117 of computer is occuring only one time in new doc. 

In [8]:
new_doc="human computer interaction computer"
new_vec=dictionary.doc2bow(new_doc.lower().split())
print(new_vec)

[(117, 2)]


so dictionary asssigns id values to all tokens and we take that id from dictionaryand then gives new vector that is count or occurrence

In [9]:
#now apply dictionary to whole text
bow_corpus=[dictionary.doc2bow(text) for text in processed_corpus]
pprint.pprint(bow_corpus)

[[(0, 1),
  (1, 1),
  (2, 1),
  (3, 1),
  (4, 1),
  (5, 1),
  (6, 1),
  (7, 1),
  (8, 1),
  (9, 1),
  (10, 1),
  (11, 1),
  (12, 1),
  (13, 1),
  (14, 1),
  (15, 1),
  (16, 1),
  (17, 1),
  (18, 1),
  (19, 1),
  (20, 1),
  (21, 1),
  (22, 1)],
 [(23, 1),
  (24, 1),
  (25, 1),
  (26, 1),
  (27, 1),
  (28, 1),
  (29, 1),
  (30, 1),
  (31, 1),
  (32, 1),
  (33, 1),
  (34, 1),
  (35, 1),
  (36, 1),
  (37, 1),
  (38, 1),
  (39, 1),
  (40, 1),
  (41, 1),
  (42, 1),
  (43, 1),
  (44, 1)],
 [(45, 1),
  (46, 2),
  (47, 1),
  (48, 1),
  (49, 1),
  (50, 1),
  (51, 1),
  (52, 1),
  (53, 1),
  (54, 1),
  (55, 1),
  (56, 1),
  (57, 1),
  (58, 1),
  (59, 1),
  (60, 1),
  (61, 1),
  (62, 1),
  (63, 1),
  (64, 1),
  (65, 1),
  (66, 1),
  (67, 1),
  (68, 1),
  (69, 1)],
 [(1, 1),
  (46, 1),
  (70, 1),
  (71, 1),
  (72, 1),
  (73, 1),
  (74, 1),
  (75, 1),
  (76, 1),
  (77, 1),
  (78, 1),
  (79, 1),
  (80, 1),
  (81, 1),
  (82, 1),
  (83, 1),
  (84, 1),
  (85, 1),
  (86, 1),
  (87, 1),
  (88, 1),
  (89, 

In [11]:
#hence we got id and no of occurrence or count above
from gensim import models
#train the model
tfidf=models.TfidfModel(bow_corpus)
#now transform the "computer autonomous" string
words="computer autonomous".lower().split()
print(tfidf[dictionary.doc2bow(words)])

[(26, 0.7071067811865476), (117, 0.7071067811865476)]


we converted words into vector using tfidf model.and this is tfidf vector value.we got id and vector.id we got using token2id and vector we got using doc2bow using count or occurrence

In [12]:
import gensim
from gensim import corpora
from pprint import pprint
from gensim.utils import simple_preprocess
import os
dict_stf=corpora.Dictionary(simple_preprocess(line) for line in open(r"sample_data.txt"))
pprint(dict_stf.token2id)

{'addressed': 52,
 'advancement': 75,
 'ai': 0,
 'algorithms': 12,
 'also': 41,
 'analytics': 30,
 'and': 13,
 'are': 14,
 'artificial': 1,
 'as': 62,
 'assist': 15,
 'automation': 16,
 'be': 53,
 'becoming': 22,
 'benefits': 63,
 'between': 67,
 'bias': 42,
 'businesses': 31,
 'but': 76,
 'collaboration': 68,
 'concerns': 43,
 'continues': 69,
 'crucial': 70,
 'customer': 32,
 'data': 44,
 'decisions': 17,
 'depends': 77,
 'deployed': 78,
 'detection': 23,
 'disease': 24,
 'displacement': 54,
 'driving': 25,
 'enabling': 18,
 'engagement': 33,
 'ensure': 64,
 'ethical': 45,
 'ethicists': 71,
 'evolve': 72,
 'for': 34,
 'forming': 55,
 'from': 2,
 'future': 79,
 'governments': 56,
 'healthcare': 3,
 'how': 80,
 'however': 46,
 'important': 47,
 'in': 19,
 'institutions': 57,
 'intelligence': 4,
 'is': 5,
 'issues': 48,
 'it': 81,
 'job': 58,
 'just': 82,
 'like': 49,
 'medicine': 26,
 'more': 35,
 'moreover': 36,
 'must': 59,
 'not': 83,
 'of': 84,
 'on': 85,
 'personalized': 27,
 'pol